dataregistry-extension/src/datatypes.ts

```diff
@@ -0,0 +1,78 @@
+/**
+ * @license BSD-3-Clause
+ *
+ * Copyright (c) 2019 Project Jupyter Contributors.
+ * Distributed under the terms of the 3-Clause BSD License.
+ */
+
+import { DataTypeNoArgs, DataTypeStringArg } from "@jupyterlab/dataregistry";
+import { INotebookModel } from "@jupyterlab/notebook";
+import { DocumentRegistry, Context } from "@jupyterlab/docregistry";
+import { Widget } from "@phosphor/widgets";
+import { Observable } from "rxjs";
+import { mimetypes }  from "./mimetypes";
+
+/**
+ * Interface describing a dataset.
+ */
+interface Dataset {
+  /**
+   * Dataset URL.
+   */
+  url: string;
+
+  /**
+   * List of child dataset identifiers.
+   */
+  children?: Array<string>;
+
+  /**
+   * Dataset label.
+   */
+  label?: string;
+
+  /**
+   * Code snippets.
+   */
+  snippets?: { [key: string]: string };
+}
+
+/**
+ * Interface for describing datasets.
+ *
+ * ## Notes
+ *
+ * -   Each dataset may consist of (and link to) other datasets.
+ */
+interface Datasets {
+  /**
+   * List of child dataset identifiers.
+   */
+  children?: Array<string>;
+
+  /**
+   * List of datasets.
+   */
+  datasets?: Array<Dataset>;
+};
+
+/**
+ * Table of data registry data types.
+ */
+const datatypes = {
+  "csv": new DataTypeStringArg<Observable<string>>(mimetypes.csv, "mimeType"), // FIXME
+  "datagrid": new DataTypeNoArgs<Observable<string>>(mimetypes.datagrid),
+  "datasetsFile": new DataTypeNoArgs<Observable<Datasets>>(mimetypes.datasetsFile),
+  "label": new DataTypeNoArgs<Observable<string>>(mimetypes.label),
+  "notebookContext": new DataTypeNoArgs<Observable<Context<INotebookModel>>>(mimetypes.notebookContext),
+  "text": new DataTypeStringArg<Observable<string>>("text/plain", "mimeType"),
+  "textContext": new DataTypeStringArg<Observable<Context<DocumentRegistry.ICodeModel>>>(mimetypes.textContext, "mimeType"),
+  "widget": new DataTypeStringArg<() => Widget>(mimetypes.widget, "label")
+};
+
+/**
+ * Exports.
+ */
+export { datatypes };
```

Comment:
I think this should just export each data type, like:

```
export const CSVDataType = new DataTypeNoArgs<Observable<string>>("text/csv");
```

We don't need to copy all these other ones. They are defined in other files.

Reply:
There may be some confusion here. This file is not defined in the `csvviewer` folder by in the parent directory.
​
In short, the idea is that, rather than having each data converter extension export various data types, thus requiring us to read individual files to see what data types are available, we should provide a single "database" of datatypes.
​
This allows us, when creating converters, to only need to consult one file, rather than many, in order to know what datatypes are implemented in the dataregistry.
​
If we have datatypes scattered across different files, we'll have to engage in an expedition to find files exporting data types and then, even supposing we codify a file in which these are defined, we have to read that separate files to know the identifier name which is exported (e.g., `CSVDataType` or `DatasetsFileDataType`, etc).
​
Instead, I'd argue that we'd be better off centralizing the declaration of data types in a single file (i.e., database). Meaning, in this refactoring, I will move the data types declarations to this `datatypes.ts` file. Then, whenever I'd want to implement a new converter, I'd only need to consult a single file.